src/backend/executor/execMain.c

```diff
@@ -2264,11 +2267,13 @@ InitPlan(QueryDesc *queryDesc, int eflags)
 			/*
 			 * A subplan will never need to do BACKWARD scan nor MARK/RESTORE.
 			 *
-			 * GPDB: We always set the REWIND flag, to delay eagerfree.
+			 * GPDB: If it's not an Init Plan, we always set the REWIND flag
+			 * to delay eagerfree.
 			 */
 			sp_eflags = eflags
 				& (EXEC_FLAG_EXPLAIN_ONLY | EXEC_FLAG_WITH_NO_DATA);
-			sp_eflags |= EXEC_FLAG_REWIND;
+			if (!bms_is_member(subplan_idx, initplans))
+				sp_eflags |= EXEC_FLAG_REWIND;
```

Comment:
Did you comapre your implementation with original implemenattion of Postgres 9.4?

```c
		 * A subplan will never need to do BACKWARD scan nor MARK/RESTORE. If
		 * it is a parameterless subplan (not initplan), we suggest that it be
		 * prepared to handle REWIND efficiently; otherwise there is no need.
		 */
		sp_eflags = eflags
			& (EXEC_FLAG_EXPLAIN_ONLY | EXEC_FLAG_WITH_NO_DATA);
		if (bms_is_member(i, plannedstmt->rewindPlanIDs))
			sp_eflags |= EXEC_FLAG_REWIND;
```

Reply:
Well, first of all we should look at gpdb [comment](delay eagerfree.):
```
 GPDB: We always set the REWIND flag, to delay eagerfree.
```
Therefore, EXEC_REWING should be set for any other SubPlan execpt the InitPlans.

Secondly, let's look at the only [usage](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/optimizer/plan/subselect.c#L990-L999) of `rewindPlanIDs`:
```
if (splan->parParam == NIL && !splan->is_initplan && !splan->useHashTable)
		root->glob->rewindPlanIDs = bms_add_member(root->glob->rewindPlanIDs,
				                           splan->plan_id);
```
In other words, some correlated SubPlans won't be added to rewindPlanIDs as well. And I guess EXEC_REWIND anyway is expected for them.

Reply:
> if (splan->parParam == NIL && !splan->is_initplan && !splan->useHashTable)
> 		root->glob->rewindPlanIDs = bms_add_member(root->glob->rewindPlanIDs,
> 				                           splan->plan_id);

rewindPlanIDs list does not contain init plans, does it?

Reply:
or you can add `Bitmapset  *initPlanIDs` member to `PlannerGlobal` struct and fill it by condition
```c
if (splan->is_initplan)
    root->glob->initPlanIDs = bms_add_member(root->glob->initPlanIDs, splan->plan_id);
```
and then check in it
```c
if (!bms_is_member(subplan_idx, plannedstmt->initPlanIDs))
    sp_eflags |= EXEC_FLAG_REWIND;
```